embassy/src/executor/raw/mod.rs

```diff
@@ -207,6 +211,148 @@ impl<F: Future + 'static> TaskStorage<F> {
+                )
+            };
+
+            return Some((guard, token));
+        }
+
+        None
+    }
+
+    fn spawn_mark_used(&self) -> bool {
+        let state = STATE_SPAWNED | STATE_RUN_QUEUED;
+        self.header
+            .state
+            .compare_exchange(0, state, Ordering::AcqRel, Ordering::Acquire)
+            .is_ok()
+    }
+
+    unsafe fn spawn_initialize(&'static self, future: *mut dyn Future<Output = T>) -> NonNull<TaskHeader> {
+        // Initialize the task
+        self.header.poll_fn.write(Self::poll);
+        self.future.write(future);
+        NonNull::new_unchecked(&self.header as *const TaskHeader as *mut TaskHeader)
+    }
+
+    unsafe fn poll(p: NonNull<TaskHeader>) {
+        let this = &*(p.as_ptr() as *const Self);
+
+        let future = Pin::new_unchecked(&mut **this.future.as_mut());
+        let waker = waker::from_task(p);
+        let mut cx = Context::from_waker(&waker);
+        match future.poll(&mut cx) {
+            Poll::Ready(res) => {
+                this.future.drop_in_place();
+                this.header.state.fetch_and(!STATE_SPAWNED, Ordering::AcqRel);
+                (*this.result.get()).write(res);
+                this.header.state.fetch_or(STATE_RESULT_PENDING, Ordering::Release);
+                this.waker.wake();
+            }
+            Poll::Pending => {}
+        }
+
+        // the compiler is emitting a virtual call for waker drop, but we know
+        // it's a noop for our waker.
+        mem::forget(waker);
+    }
+}
+
+unsafe impl<F> Sync for ScopedTaskStorage<F> {}
+
+/// TODO
+pub struct SpawnScopedGuard<'a, T> {
+    storage: &'a ScopedTaskStorage<T>,
+}
+
+impl<'a, T> Future for SpawnScopedGuard<'a, T> {
+    type Output = T;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.storage.waker.register(cx.waker());
+        if self
+            .storage
+            .header
+            .state
+            .fetch_and(!STATE_RESULT_PENDING, Ordering::Acquire)
+            & STATE_RESULT_PENDING
+            != 0
+        {
+            // Safety: STATE_RESULT_PENDING ensures that result is init
+            Poll::Ready(unsafe { (*self.storage.result.get()).assume_init_read() })
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+impl<'a, T> Drop for SpawnScopedGuard<'a, T> {
+    fn drop(&mut self) {
+        if self.storage.header.state.load(Ordering::Relaxed) & STATE_SPAWNED != 0 {
+            panic!("SpawnScopedGuard dropped while task is still queued! Consider awaiting the guard")
+        }
+
+        if self
+            .storage
+            .header
+            .state
+            .fetch_and(!STATE_RESULT_PENDING, Ordering::Acquire)
+            & STATE_RESULT_PENDING
+            != 0
+        {
+            // Safety: STATE_RESULT_PENDING ensures that result is init
+            unsafe {
+                (*self.storage.result.get()).assume_init_drop();
```

Comment:
you should first drop, then clear the flag. By clearing the flag you give permission to other threads to overwrite the task storage, which will race with us doing the drop.

Reply:
My idea was to protect this by borrowing `ScopedTaskStorage` for the duration of the guard, which would prevent any new spawn until guard is dropped and the only write is in the poll method with `Ordering::Release`. But now I see that I needed a `mut` borrow to make it exclusive, my bad.

To make it thread safe without exclusive borrow I think you would need a CAS-like mutex, which is not available on some MCUs. But I see that executor already has some CAS operations, so I guess it's not a problem?

Reply:
this shouldn't need CAS, all you need to do is to clear `STATE_RESULT_PENDING` after dropping, not before.

(And ordering should be `Release` I think, to prevent the drop to be moved down past it)